markovflow/sde/sde_utils.py

```diff
@@ -226,3 +242,219 @@ def func(x, t=None, A=A, b=b):
+        2. func_q and func_p should support batching as quadrature points are passed to it.
+        3. The function only supports state-dim=1 currently.
+
+    """
+
+    n_transitions = ssm_p_process_covar.shape[0]
+    tf.debugging.assert_equal(
+        ssm_q_process_covar.get_shape(), tf.TensorShape([n_transitions, 1, 1])
+    )
+    tf.debugging.assert_equal(
+        ssm_p_process_covar.get_shape(), tf.TensorShape([n_transitions, 1, 1])
+    )
+    tf.debugging.assert_equal(
+        ssm_q_marginals_mean.get_shape(), tf.TensorShape([n_transitions + 1, 1])
+    )
+    tf.debugging.assert_equal(
+        ssm_q_marginals_covar_diag.get_shape(), tf.TensorShape([n_transitions + 1, 1])
+    )
+
+    ssm_p_process_covar_inv = tf.linalg.cholesky_solve(
+        tf.linalg.cholesky(ssm_p_process_covar),
+        tf.eye(
+            ssm_p_process_covar.shape[-1],
+            dtype=ssm_p_process_covar.dtype,
+            batch_shape=[ssm_p_process_covar.shape[0]],
+        ),
+    )
+
+    C_log_det = tf.linalg.logdet(ssm_q_process_covar) - tf.linalg.logdet(ssm_p_process_covar)
+    C_term_trace = tf.linalg.trace(ssm_p_process_covar_inv * ssm_q_process_covar)
+    C_term_D = ssm_q_marginals_mean.shape[-1]
+
+    C = - C_log_det - C_term_D + C_term_trace
+
+    def func(x):
+        func_q_val = func_q(x=x)  # [n_pnts, num_transitions, 1]
+        func_p_val = func_p(x=x)  # [n_pnts, num_transitions, 1]
+
+        tf.debugging.assert_shapes(
+            [
+                (func_q_val, ("B", "N", "D")),
+                (func_p_val, ("B", "N", "D")),
+                (ssm_p_process_covar_inv[:, :, 0], ("N", "D")),
+            ]
+        )
+
+        val = (func_p_val - func_q_val) * (func_p_val - func_q_val) * ssm_p_process_covar_inv[:, :, 0]
+
+        return val
+
+    m = ssm_q_marginals_mean[:-1]
+    S = ssm_q_marginals_covar_diag[:-1]
+
+    fn_difference = NDiagGHQuadrature(ssm_q_marginals_mean.shape[-1], quadrature_pnts)(func, mean=m, var=S)
+
+    vals = fn_difference + C[..., None]
+
+    tf.debugging.assert_shapes(
+        [
+            (m, ("N", "D")),
+            (vals, ("N", "D")),
+        ]
+    )
+
+    vals = 0.5 * tf.reduce_sum(vals)
+    return vals
+
+
+def ssm_to_btd_nat(ssm: StateSpaceModel) -> BTDGaussian:
+    """
+    Get natural parameters from a SSM model as a BTDGaussian.
+    """
+    assert ssm.batch_shape.is_compatible_with(
+        tf.TensorShape([])), "ssm_to_btd_nat function currently does not support batch!"
+
+    nat1, nat_diag, nat_subdiag = ssm_to_naturals(ssm)
+
+    nat_btd = BTDGaussian(nat1, nat_diag, nat_subdiag)
+
+    return nat_btd
+
+
+def SSM_KL_with_grads_wrt_exp_params(
+        ssm_q: StateSpaceModel, ssm_p: StateSpaceModel
+) -> [tf.Tensor, [tf.Tensor, tf.Tensor, tf.Tensor]]:
+    """
+    The function calculates the KL divergence between two SSMs, KL[SSM-q || SSM-p], and returns the gradients wrt the
+    expectation parameters of SSM-q.
```

Comment:
It would be good to have some of the maths here to understand the code, and where the gradient flows